File: linux/cmd/abacad/connect.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"

	"abacad-linux/internal/x11"
)

// runConnect implements `abacad connect`: the device-authorization grant (RFC
// 8628) that enrolls this machine without copy-pasting a token. It asks the
// server for a short code, prints the URL to approve it, polls until the human
// approves in their browser, then writes the issued token to the config file so
// a plain `abacad` afterwards just runs.
func runConnect(args []string) {
	fs := flag.NewFlagSet("connect", flag.ExitOnError)
	server := fs.String("server", "https://abacad.ai", "abacad server base URL")
	_ = fs.Parse(args)

	base := strings.TrimRight(strings.TrimSpace(*server), "/")
	if base == "" {
		log.Fatal("empty --server")
	}

	platform := detectPlatform()
	log.Printf("this device: %s", platform)

	start, err := pairStart(base, platform)
	if err != nil {
		log.Fatalf("start pairing: %v", err)
	}

	link := start.VerificationURIComplete
	if link == "" {
		link = start.VerificationURI
	}
	fmt.Fprintf(os.Stderr, "\nTo connect this device, open:\n\n    %s\n\n", link)
	fmt.Fprintf(os.Stderr, "and approve it while signed in (code: %s). Waiting…\n", start.UserCode)

	tok, err := pairPoll(base, start)
	if err != nil {
		log.Fatalf("pairing: %v", err)
	}

	serverURL, token, err := splitToken(tok.WssURL)
	if err != nil {
		log.Fatalf("parse issued URL: %v", err)
	}
	path, err := saveConfig(map[string]string{"server_url": serverURL, "token": token})
	if err != nil {
		log.Fatalf("save config: %v", err)
	}

	fmt.Fprintf(os.Stderr, "\n✓ Approved. Saved credentials to %s\n", path)
	fmt.Fprintf(os.Stderr, "  Start the device now with:  abacad\n")
}

// detectPlatform reports "linux" when an X server is reachable and
// "linux-headless" otherwise, so a display-less box advertises shell-only and
// the agent won't try to screenshot a rack server.
func detectPlatform() string {
	if x, err := x11.Open(); err == nil {
		x.Close()
		return "linux"
	}
	return "linux-headless"
}

type startResp struct {
	DeviceCode              string `json:"device_code"`
	UserCode                string `json:"user_code"`
	VerificationURI         string `json:"verification_uri"`
	VerificationURIComplete string `json:"verification_uri_complete"`
	Interval                int    `json:"interval"`
	ExpiresIn               int    `json:"expires_in"`
}

type pollResp struct {
	Status      string `json:"status"`
	DeviceToken string `json:"device_token"`
	WssURL      string `json:"wss_url"`
	Error       string `json:"error"`
	Interval    int    `json:"interval"`
}

func pairStart(base, platform string) (*startResp, error) {
	body, _ := json.Marshal(map[string]string{"platform": platform})
	resp, err := http.Post(base+"/api/devices/pair/start", "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("server said %s: %s", resp.Status, serverError(data))
	}
	var s startResp
	if err := json.Unmarshal(data, &s); err != nil {
		return nil, err
	}
	if s.DeviceCode == "" {
		return nil, fmt.Errorf("server returned no device_code")
	}
	return &s, nil
}

// pairPoll waits for approval, honoring the server's interval hint and giving up
// after the pairing's stated lifetime.
func pairPoll(base string, start *startResp) (*pollResp, error) {
	interval := time.Duration(start.Interval) * time.Second
	if interval < time.Second {
		interval = 3 * time.Second
	}
	deadline := time.Now().Add(time.Duration(max(start.ExpiresIn, 60)) * time.Second)
	body, _ := json.Marshal(map[string]string{"device_code": start.DeviceCode})

	for {
		if time.Now().After(deadline) {
			return nil, fmt.Errorf("timed out waiting for approval")
		}
		resp, err := http.Post(base+"/api/devices/pair/poll", "application/json", bytes.NewReader(body))
		if err != nil {
			return nil, err
		}
		data, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
		resp.Body.Close()

		switch resp.StatusCode {
		case http.StatusOK:
			var p pollResp
			if err := json.Unmarshal(data, &p); err != nil {
				return nil, err
			}
			if p.WssURL == "" {
				return nil, fmt.Errorf("server approved but returned no wss_url")
			}
			return &p, nil
		case http.StatusAccepted: // still pending
			time.Sleep(interval)
		default: // 403 denied / 404 unknown / 410 expired-or-used → terminal
			return nil, fmt.Errorf("%s", serverError(data))
		}
	}
}

// splitToken separates the issued wss URL into the token-free base URL and the
// token, matching the config's split server_url / token keys and keeping the
// secret out of the URL the daemon dials.
func splitToken(wssURL string) (base, token string, err error) {
	u, err := url.Parse(wssURL)
	if err != nil {
		return "", "", err
	}
	q := u.Query()
	token = q.Get("token")
	q.Del("token")
	u.RawQuery = q.Encode()
	return u.String(), token, nil
}

// saveConfig writes ~/.config/abacad/config with 0600 perms (the token is a
// bearer secret; there is no keychain on a headless box, so a private file is
// the right store — same trust model as an SSH key).
func saveConfig(kv map[string]string) (string, error) {
	dir, err := os.UserConfigDir()
	if err != nil {
		return "", err
	}
	d := filepath.Join(dir, "abacad")
	if err := os.MkdirAll(d, 0o700); err != nil {
		return "", err
	}
	path := filepath.Join(d, "config")
	var b strings.Builder
	b.WriteString("# Written by `abacad connect`. Re-run it to replace these.\n")
	for _, k := range []string{"server_url", "token"} {
		if v := kv[k]; v != "" {
			fmt.Fprintf(&b, "%s = %s\n", k, v)
		}
	}
	if err := os.WriteFile(path, []byte(b.String()), 0o600); err != nil {
		return "", err
	}
	return path, nil
}

// serverError pulls the {"error": …} message out of a JSON error body, falling
// back to the raw text.
func serverError(data []byte) string {
	var e struct {
		Error string `json:"error"`
	}
	if json.Unmarshal(data, &e) == nil && e.Error != "" {
		return e.Error
	}
	return strings.TrimSpace(string(data))
}

File: macos/Sources/abacad/Connect.swift
import Foundation

// `abacad connect` — the device-authorization grant (RFC 8628) that enrolls this
// Mac without pasting a token into the menu-bar panel. It asks the server for a
// short code, prints the URL to approve it, polls until the human approves in
// their browser, then stores the issued wss://…?token=… URL via Prefs (login
// Keychain) so the menu-bar app auto-connects on next launch. The CLI peer of the
// Linux and Windows `abacad connect`.
enum ConnectFlow {
    static let defaultServer = "https://abacad.ai"

    /// Entry from the process launcher. Returns a process exit code.
    static func run(_ args: [String]) -> Int32 {
        // Stream progress even when stdout is piped/redirected: the URL + code
        // must appear before we block polling, but stdout is block-buffered off a
        // TTY, which would hide them until exit.
        setvbuf(stdout, nil, _IONBF, 0)

        var server = defaultServer
        var i = 1 // args[0] is "connect"
        while i < args.count {
            let a = args[i]
            if a == "--server" || a == "-s", i + 1 < args.count {
                server = args[i + 1]
                i += 2
                continue
            }
            if a.hasPrefix("--server=") {
                server = String(a.dropFirst("--server=".count))
            }
            i += 1
        }
        server = server.trimmingCharacters(in: .whitespaces)
        while server.hasSuffix("/") { server.removeLast() }
        if server.isEmpty {
            printErr("empty --server")
            return 1
        }

        // Bridge the async flow to this synchronous entry: a class holder keeps the
        // result Sendable-safe across the Task boundary.
        final class Box { var code: Int32 = 1 }
        let box = Box()
        let done = DispatchSemaphore(value: 0)
        Task {
            box.code = await runAsync(server: server)
            done.signal()
        }
        done.wait()
        return box.code
    }

    private static func runAsync(server: String) async -> Int32 {
        // 1. Start the pairing, reporting our platform so the approval page shows it.
        guard let s = await post(server + "/api/devices/pair/start", body: ["platform": "macos"]) else {
            printErr("start pairing: could not reach \(server)")
            return 1
        }
        guard s.status == 201 || s.status == 200,
              let start = s.json, !start.string("device_code").isEmpty
        else {
            printErr("start pairing: " + serverError(s.json, s.status))
            return 1
        }
        let deviceCode = start.string("device_code")
        let userCode = start.string("user_code")
        var link = start.string("verification_uri_complete")
        if link.isEmpty { link = start.string("verification_uri") }
        let interval = max(start.int("interval"), 1)
        let expiresIn = max(start.int("expires_in"), 60)

        print("")
        print("To connect this device, open:")
        print("")
        print("    \(link)")
        print("")
        print("and approve it while signed in (code: \(userCode)). Waiting…")

        // 2. Poll until the human approves, honoring the interval + lifetime hints.
        let deadline = Date().addingTimeInterval(TimeInterval(expiresIn))
        while true {
            if Date() > deadline {
                printErr("timed out waiting for approval")
                return 1
            }
            guard let p = await post(server + "/api/devices/pair/poll", body: ["device_code": deviceCode]) else {
                printErr("poll: could not reach \(server)")
                return 1
            }
            switch p.status {
            case 200:
                let wss = p.json?.string("wss_url") ?? ""
                let token = p.json?.string("device_token") ?? ""
                if wss.isEmpty {
                    printErr("approved but server returned no wss_url")
                    return 1
                }
                // The server's wss_url already carries ?token=; store it whole (the
                // format Prefs/WebSocketClient expect), re-attaching only if absent.
                Prefs.serverURL = withToken(wss, token)
                print("")
                print("✓ Approved. Credentials saved to the login Keychain.")
                print("  Launch abacad (the menu-bar app) to go online — it auto-connects.")
                return 0
            case 202: // still pending
                try? await Task.sleep(nanoseconds: UInt64(interval) * 1_000_000_000)
            default: // 403 denied / 404 unknown / 410 expired-or-used → terminal
                printErr(serverError(p.json, p.status))
                return 1
            }
        }
    }

    /// POST a JSON body and return the status + parsed object, or nil on a
    /// transport error. Uses the loose Json helpers to match the client's style.
    private static func post(_ urlString: String, body: [String: Any]) async -> (status: Int, json: [String: Any]?)? {
        guard let url = URL(string: urlString) else { return nil }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = Data(Json.string(body).utf8)
        guard let (data, resp) = try? await URLSession.shared.data(for: req) else { return nil }
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        let text = String(data: data, encoding: .utf8) ?? ""
        return (status, Json.object(text))
    }

    // withToken keeps the token in the URL (Prefs stores one combined string that
    // WebSocketClient splits at dial time). The server already appends ?token=, so
    // this only adds it in the unlikely case it's missing.
    private static func withToken(_ wss: String, _ token: String) -> String {
        if token.isEmpty || wss.contains("token=") { return wss }
        let sep = wss.contains("?") ? "&" : "?"
        let enc = token.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? token
        return wss + sep + "token=" + enc
    }

    private static func serverError(_ json: [String: Any]?, _ status: Int) -> String {
        let msg = json?.string("error") ?? ""
        return msg.isEmpty ? "server said \(status)" : msg
    }

    private static func printErr(_ s: String) {
        FileHandle.standardError.write(Data((s + "\n").utf8))
    }
}
